orttraining/orttraining/test/python/orttraining_test_ortmodule_api.py

```diff
@@ -3131,3 +3132,131 @@ def test_debug_options_log_level_validation_fails_on_type_mismatch():
+            b = batch['b']
+            a = batch['a'] if batch['a'] else torch.FloatTensor([2.0]).cuda()
+            return self.dummy + a + b
+
+    device = 'cuda'
+    N, D_in, H, D_out = 64, 784, 500, 10
+    pt_model = DictNet().to(device)
+    ort_model = ORTModule(copy.deepcopy(pt_model))
+    x = {'a': None, 'b': torch.randn(N, D_in, device=device)}
+    y = copy.deepcopy(x)
+
+    _test_helpers.assert_values_are_close(pt_model(x), ort_model(y))
+
+def test_ortmodule_dict_input_with_nested_values():
+    class DictNet(torch.nn.Module):
+        def __init__(self):
+            super(DictNet, self).__init__()
+            self.dummy = torch.nn.Parameter(torch.FloatTensor([0]))
+
+        def forward(self, batch):
+            a = batch['one_value']
+            b = batch['two_value']['three_value']
+            c = batch['two_value']['four_value']
+            d = batch['five_value']['six_value']
+            e = batch['five_value']['seven_value']['eight_value']
+            return self.dummy + a + b + c + d + e
+
+    device = 'cuda'
+    N, D_in, H, D_out = 64, 784, 500, 10
+    pt_model = DictNet().to(device)
+    ort_model = ORTModule(copy.deepcopy(pt_model))
+    x = {
+            'one_value': torch.randn(N, D_in, device=device),
+            'two_value': {
+                'three_value': torch.randn(N, D_in, device=device),
+                'four_value': torch.randn(N, D_in, device=device)
+            },
+            'five_value': {
+                'six_value': torch.randn(N, D_in, device=device),
+                'seven_value': {
+                    'eight_value': torch.randn(N, D_in, device=device)
+                }
+            }
+        }
+    y = copy.deepcopy(x)
+
+    _test_helpers.assert_values_are_close(pt_model(x), ort_model(y))
+
+
+def test_ortmodule_list_dict_input_with_nested_values():
+    class ListDictNet(torch.nn.Module):
+        def __init__(self):
+            super(ListDictNet, self).__init__()
+            self.dummy = torch.nn.Parameter(torch.FloatTensor([3]))
+
+        def forward(self, batch):
+            a = batch['one_value'][0]
+            b = batch['two_value'][0]
+            c = batch['two_value'][1]
+            d = batch['three_value'][0]
+            e = batch['three_value'][1]['four_value']
+            return self.dummy + a + b + c + d + e
+
+    device = 'cuda'
+    N, D_in, H, D_out = 64, 784, 500, 10
+    pt_model = ListDictNet().to(device)
+    ort_model = ORTModule(copy.deepcopy(pt_model))
+    x = {
+            'one_value': [torch.randn(N, D_in, device=device)],
+            'two_value': [torch.randn(N, D_in, device=device), torch.randn(N, D_in, device=device)],
+            'three_value': [
+                torch.randn(N, D_in, device=device),
+                {
+                    'four_value': torch.randn(N, D_in, device=device)
+                }
+            ]
+        }
+    y = copy.deepcopy(x)
+
+    _test_helpers.assert_values_are_close(pt_model(x), ort_model(y))
```

Comment:
should we add test with registered buffers and kwargs side by side with dict?

Reply:
Thanks for this suggestion. I uncovered a bug where the ```nested``` input and ```kwargs``` did not work well together. The problem was that upon expanding the nested input, the ```num_positionals_non_none``` was no longer valid since the ```input_names``` were expanded. So, I had to adjust the ```num_positionals_non_none``` to be ```num_expanded_positionals_non_none```. This unit test helped identify that problem.